snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/MrcmTypeRequest.java

```diff
@@ -0,0 +1,160 @@
+ * limitations under the License.
+ */
+package com.b2international.snowowl.snomed.datastore.request;
+
+import static com.b2international.index.revision.Revision.Fields.ID;
+import static com.b2international.snowowl.snomed.common.SnomedConstants.Concepts.CONCEPT_MODEL_DATA_ATTRIBUTE;
+import static com.b2international.snowowl.snomed.common.SnomedConstants.Concepts.CONCEPT_MODEL_OBJECT_ATTRIBUTE;
+import static com.b2international.snowowl.snomed.datastore.index.entry.SnomedRefSetMemberIndexEntry.Fields.MRCM_RULE_REFSET_ID;
+import static com.google.common.base.Strings.nullToEmpty;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.b2international.commons.exceptions.BadRequestException;
+import com.b2international.commons.options.Options;
+import com.b2international.snowowl.core.domain.BranchContext;
+import com.b2international.snowowl.core.events.Request;
+import com.b2international.snowowl.snomed.common.SnomedRf2Headers;
+import com.b2international.snowowl.snomed.core.domain.SnomedConcept;
+import com.b2international.snowowl.snomed.core.domain.refset.SnomedRefSetType;
+import com.b2international.snowowl.snomed.core.domain.refset.SnomedReferenceSetMembers;
+import com.b2international.snowowl.snomed.datastore.index.entry.SnomedConceptDocument;
+
+/**
+ * @since 8.7.0
+ */
+public class MrcmTypeRequest implements Request<BranchContext, SnomedReferenceSetMembers> {
+	
+	private static final long serialVersionUID = 1L;
+
+	public enum ATTRIBUTE_TYPE {
+		DATA,
+		OBJECT,
+		ALL;
+		
+		public static ATTRIBUTE_TYPE getByNameIgnoreCase(String name) {
+			for (final ATTRIBUTE_TYPE type : values()) {
+				if (nullToEmpty(name).equalsIgnoreCase(type.toString())) {
+					return type;
+				}
+			}
+			throw new BadRequestException("Unknown attribute type: '%s'", name);
+		}
+	}
+	
+	private Set<String> selfIds;
+	private Set<String> ruleParentIds; 
+	private Set<String> refSetIds;
+	private List<String> moduleIds;
+	private ATTRIBUTE_TYPE attributeType;
+	
+	public void setSelfIds(Set<String> selfIds) {
+		this.selfIds = selfIds;
+	}
+
+	public void setParentIds(Set<String> ruleParentIds) {
+		this.ruleParentIds = ruleParentIds;
+	}
+
+	public void setRefSetIds(Set<String> refSetIds) {
+		this.refSetIds = refSetIds;
+	}
+
+	public void setModuleIds(List<String> moduleIds) {
+		this.moduleIds = moduleIds;
+	}
+
+	public void setAttributeType(ATTRIBUTE_TYPE attributeType) {
+		this.attributeType = attributeType;
+	}
+	
+	@Override
+	public SnomedReferenceSetMembers execute(BranchContext context) {
+		Set<String> inScopeRefSetIds = SnomedRequests.prepareSearchMember()
+				.all()
+				.filterByActive(true)
+				.filterByRefSetType(SnomedRefSetType.MRCM_MODULE_SCOPE)
+				.filterByReferencedComponent(moduleIds)
+				.setFields(ID, MRCM_RULE_REFSET_ID)
+				.build()
+				.execute(context)
+				.stream()
```

Comment:
Please don't use all() anymore unless you have a really good reason to do so. Also, using all + stream together is a no go. Please set the limit correctly.

Reply:
This request queries a small set of module scope members, which should be kept in practice at a low member count (the current tally is 3, and the expected count is 3 × number of modules declaring MRCM rules). `stream()` is called on the returned collection resource and processes individual members, not batches.